static/js/formatters-internal.js

```diff
@@ -311,164 +311,117 @@ export function image(simpleOrComplexImage = {}, desiredSize = '200x', atLeastAs
+    }
 
-  const dynamicUrl = isEuImage
-    ? _getEuImageDynamicUrl(image, desiredSize, atLeastAsLarge)
-    : _getUsImageDynamicUrl(image.url, desiredSize, atLeastAsLarge);
+    const hostname = url.hostname.replace(/^[a-z]+\./, 'dyn.');
+    const formatOptionsString = _getImageFormatOptions(desiredSize, atLeastAsLarge, image.width, image.height);
 
-  return Object.assign(
-    {},
-    image,
-    {
-      url: dynamicUrl.replace('http://', 'https://')
-    }
-  );
+    return Object.assign(
+      {},
+      image,
+      {
+        url: `https://${hostname}${urlPath}${formatOptionsString}`,
+      }
+    );
+
+  } catch (error) {
+    throw new Error(`Error processing image url ${image.url}: ${error}`);
+  }
 }
 
 /**
- * Given a US image url, returns the dynamic url.
+ * Construct the format options string with given parameters.
  *
- * @param {string} imageUrl Image's url (e.g.
- *   'https://dynl.mktgcdn.com/p/ldMLwj1JkN94-2pwh6CjR_OMy4KnexHJCfZhPAZCbi0/196x400.jpg',
- *   'https://a.mktgcdn.com/p/ldMLwj1JkN94-2pwh6CjR_OMy4KnexHJCfZhPAZCbi0/196x400.jpg')
  * @param {string} desiredSize The desired size of the image ('<width>x<height>')
  * @param {boolean} atLeastAsLarge Whether the image should be at least as large as the desired
  *                                 size in one dimension or smaller than the desired size in both
  *                                 dimensions.
- * @returns {string} A dynamic url (e.g. 'https://dynl.mktgcdn.com/p/ldMLwj1JkN94-2pwh6CjR_OMy4KnexHJCfZhPAZCbi0/200x1.jpg')
+ * @param {number?} fullSizeWidth The full size width of the original image
+ * @param {number?} fullSizeHeight The full size height of the original image
+ * @returns {string} A string representing the format options (e.g. 'height=200,width=100,fit=cover')
  */
-function _getUsImageDynamicUrl(imageUrl, desiredSize, atLeastAsLarge) {
-  const [urlWithoutExtension, extension] = _splitStringOnIndex(imageUrl, imageUrl.lastIndexOf('.'));
-  const [urlBeforeDimensions, dimensions] = _splitStringOnIndex(urlWithoutExtension, urlWithoutExtension.lastIndexOf('/') + 1);
-  const fullSizeDims = dimensions.split('x');
-
-  let desiredWidth, desiredHeight;
+function _getImageFormatOptions(desiredSize, atLeastAsLarge, fullSizeWidth, fullSizeHeight) {
   let desiredDims = desiredSize.split('x');
+  const desiredWidthProvided = desiredDims[0] !== '' && desiredDims[0] !== '1';
+  const desiredHeightProvided = desiredDims[1] !== '' && desiredDims[1] !== '1';
 
-  if (desiredDims[0] !== '') {
+  // both dimensions are not provided, return original image
+  if (!desiredWidthProvided && !desiredHeightProvided) {
+    return '';
+  }
+
+  const originalRatio =
+    (!!fullSizeWidth && !!fullSizeHeight) ? (fullSizeWidth / fullSizeHeight) : undefined;
+  let desiredWidth;
+  let desiredHeight;
+  let formatOptions = ['fit=contain'];
+  if (desiredWidthProvided) {
     desiredWidth = Number.parseInt(desiredDims[0]);
```

Comment:
why do we need both a desiredWidthProvided and a desiredWidth?

Reply:
if this is a boolean, can the name start with `is` or `has` (e.g. `hasDesiredWidth` or `isDesiredWidthProvided`)?